File: modules/apps/desktop/internal/adapter/webui/creating_test.go
package webui_test

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"connectrpc.com/connect"

	v1 "github.com/jiva-studio/numen/modules/libs/protocol/gen/numen/v1"
)

// The window makes a note from the picture: a name, and the one link that puts
// it where the person reached out from.

func fileAt(t *testing.T, root, path string) string {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(root, filepath.FromSlash(path)))
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	return string(raw)
}

// TestANoteIsMadeCarryingTheLinkThatSeatsIt. A note made in a seat of another
// says so itself: the link is in the note that was made, so one write leaves it
// joined.
func TestANoteIsMadeCarryingTheLinkThatSeatsIt(t *testing.T) {
	f := quitting(t, nil, map[string]string{
		"Ontology.md": "---\ntitle: Ontology\n---\n\n# Ontology\n",
	})

	answer, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title: "Entropy",
		Links: []*v1.NewLink{{To: "Ontology.md", Seat: v1.Seat_SEAT_PARENT}},
	}))
	if err != nil {
		t.Fatal(err)
	}
	if refusal := answer.Msg.GetRefusal(); refusal != v1.Refusal_REFUSAL_UNSPECIFIED {
		t.Fatalf("the note was refused: %v", refusal)
	}
	if path := answer.Msg.GetPath(); path != "Entropy.md" {
		t.Fatalf("the note was filed at %q", path)
	}

	made := fileAt(t, f.root, "Entropy.md")
	if !strings.Contains(made, "to: Ontology") {
		t.Errorf("the note carries no link to the note it was made from:\n%s", made)
	}
	if !strings.Contains(made, "role: parent") {
		t.Errorf("the link does not seat the note it was made from as its parent:\n%s", made)
	}
	if !strings.Contains(made, "id: ") {
		t.Errorf("a note the application made carries no identifier:\n%s", made)
	}
}

// TestANoteMadeInTheChildSeatIsTheParentsChild. The seat travels as the link the
// new note writes, so the role is the one that puts each of the two where the
// person put them.
func TestANoteMadeInTheChildSeatIsTheParentsChild(t *testing.T) {
	f := quitting(t, nil, map[string]string{
		"Ontology.md": "---\ntitle: Ontology\n---\n\n# Ontology\n",
	})

	answer, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title: "Entropy",
		Links: []*v1.NewLink{{To: "Ontology.md", Seat: v1.Seat_SEAT_CHILD, Label: "follows from"}},
	}))
	if err != nil {
		t.Fatal(err)
	}

	made := fileAt(t, f.root, answer.Msg.GetPath())
	if !strings.Contains(made, "role: child") {
		t.Errorf("the link does not seat the note it was made from as its child:\n%s", made)
	}
	if !strings.Contains(made, "label: follows from") {
		t.Errorf("what the person wrote on the link is not in the note:\n%s", made)
	}
}

// TestANoteIsMadeInTheFolderItWasAskedFor. The person's arrangement of their own
// folders is followed and never altered.
func TestANoteIsMadeInTheFolderItWasAskedFor(t *testing.T) {
	f := quitting(t, nil, map[string]string{
		"physics/Ontology.md": "---\ntitle: Ontology\n---\n\n# Ontology\n",
	})

	answer, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title:  "Entropy",
		Folder: "physics",
		Links:  []*v1.NewLink{{To: "physics/Ontology.md", Seat: v1.Seat_SEAT_PARENT}},
	}))
	if err != nil {
		t.Fatal(err)
	}
	if path := answer.Msg.GetPath(); path != "physics/Entropy.md" {
		t.Fatalf("the note was filed at %q", path)
	}
	// The link is written by the other note's name, which is what carries across
	// folders.
	if made := fileAt(t, f.root, "physics/Entropy.md"); !strings.Contains(made, "to: Ontology") {
		t.Errorf("the link was not written by name:\n%s", made)
	}
}

// TestANoteMadeIsInTheIndexBeforeTheAnswerComesBack. This is what puts a note
// in the picture as soon as it exists.
func TestANoteMadeIsInTheIndexBeforeTheAnswerComesBack(t *testing.T) {
	f := quitting(t, nil, map[string]string{
		"Ontology.md": "---\ntitle: Ontology\n---\n\n# Ontology\n",
	})

	if _, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title: "Entropy",
		Links: []*v1.NewLink{{To: "Ontology.md", Seat: v1.Seat_SEAT_PARENT}},
	})); err != nil {
		t.Fatal(err)
	}

	found, err := f.opened.Index.Queries().Notes(t.Context(), f.opened.Vault.ID, []string{"Entropy.md"})
	if err != nil {
		t.Fatal(err)
	}
	if found["Entropy.md"].Title != "Entropy" {
		t.Errorf("the index answers %q for a note that was just made", found["Entropy.md"].Title)
	}
}

// TestANoteMadeWhereOneAlreadyIsIsRefused. Whether the path was free is the
// filesystem's to answer at the moment the file is made, and what is there is
// left as it stands.
func TestANoteMadeWhereOneAlreadyIsIsRefused(t *testing.T) {
	const held = "---\ntitle: Ontology\n---\n\n# Ontology\n"
	f := quitting(t, nil, map[string]string{"Ontology.md": held})

	answer, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title: "Ontology",
	}))
	if err != nil {
		t.Fatal(err)
	}
	if refusal := answer.Msg.GetRefusal(); refusal != v1.Refusal_REFUSAL_OCCUPIED {
		t.Errorf("a name already taken was answered with %v", refusal)
	}
	if answer.Msg.GetPath() != "" {
		t.Errorf("a note that was not made was filed at %q", answer.Msg.GetPath())
	}
	if now := fileAt(t, f.root, "Ontology.md"); now != held {
		t.Errorf("the note that was already there was written:\n%s", now)
	}
}

// TestASeatNoLinkWritesLeavesNoNote. A sibling is another child of a shared
// parent, so nothing writes one, and the refusal comes before the file does.
func TestASeatNoLinkWritesLeavesNoNote(t *testing.T) {
	f := quitting(t, nil, map[string]string{
		"Ontology.md": "---\ntitle: Ontology\n---\n\n# Ontology\n",
	})

	_, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title: "Entropy",
		Links: []*v1.NewLink{{To: "Ontology.md", Seat: v1.Seat_SEAT_SIBLING}},
	}))
	if connect.CodeOf(err) != connect.CodeInvalidArgument {
		t.Fatalf("a sibling was answered with %v", err)
	}
	if _, err := os.Stat(filepath.Join(f.root, "Entropy.md")); err == nil {
		t.Error("a note was made for a link that cannot be written")
	}
}

// TestTwoNotesAreJoinedFromTheOneTheLinkIsWrittenIn. A link is one end's
// account of a relationship: the other note is not touched.
func TestTwoNotesAreJoinedFromTheOneTheLinkIsWrittenIn(t *testing.T) {
	const other = "---\ntitle: Entropy\n---\n\n# Entropy\n"
	f := quitting(t, nil, map[string]string{
		"Ontology.md": "---\ntitle: Ontology\ntags:\n  - draft\n---\n\n# Ontology\n",
		"Entropy.md":  other,
	})

	answer, err := f.client.Join(t.Context(), connect.NewRequest(&v1.JoinRequest{
		Path: "Ontology.md",
		Link: &v1.NewLink{To: "Entropy.md", Seat: v1.Seat_SEAT_CHILD},
	}))
	if err != nil {
		t.Fatal(err)
	}
	if refusal := answer.Msg.GetRefusal(); refusal != v1.Refusal_REFUSAL_UNSPECIFIED {
		t.Fatalf("the link was refused: %v", refusal)
	}

	joined := fileAt(t, f.root, "Ontology.md")
	if !strings.Contains(joined, "to: Entropy") || !strings.Contains(joined, "role: child") {
		t.Errorf("the note does not say what it was joined to:\n%s", joined)
	}
	// The keys the person wrote are theirs, and come out as they went in.
	if !strings.Contains(joined, "tags:\n  - draft\n") {
		t.Errorf("a key the application does not own was rewritten:\n%s", joined)
	}
	if now := fileAt(t, f.root, "Entropy.md"); now != other {
		t.Errorf("the note at the other end was written:\n%s", now)
	}
}

// TestANoteIsMadeUnderTheNoteItWasMadeFromAndNotItsNamesake. A name is read
// back as an exact path from the root before it is read as a neighbour, so a
// note made from one of two notes sharing a name carries the path it was made
// from. Written by name, it would hang off the other one, and the picture the
// person is looking at would not draw it at all.
func TestANoteIsMadeUnderTheNoteItWasMadeFromAndNotItsNamesake(t *testing.T) {
	f := quitting(t, nil, nil)
	// Both are made through the window, which is what puts them in the index:
	// the vault is scanned at startup, and a test is not started.
	for _, folder := range []string{"", "physics"} {
		if _, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
			Title: "Ontology", Folder: folder,
		})); err != nil {
			t.Fatal(err)
		}
	}

	answer, err := f.client.Create(t.Context(), connect.NewRequest(&v1.CreateRequest{
		Title:  "Entropy",
		Folder: "physics",
		Links:  []*v1.NewLink{{To: "physics/Ontology.md", Seat: v1.Seat_SEAT_PARENT}},
	}))
	if err != nil {
		t.Fatal(err)
	}

	if made := fileAt(t, f.root, answer.Msg.GetPath()); !strings.Contains(made, "to: physics/Ontology") {
		t.Errorf("the link was written by a name two notes answer to:\n%s", made)
	}

	around, err := f.client.Neighbourhood(t.Context(),
		connect.NewRequest(&v1.NeighbourhoodRequest{Path: "physics/Ontology.md"}))
	if err != nil {
		t.Fatal(err)
	}
	var children []string
	for _, related := range around.Msg.GetRelated() {
		if related.GetSeat() == v1.Seat_SEAT_CHILD {
			children = append(children, related.GetNote().GetPath())
		}
	}
	if len(children) != 1 || children[0] != "physics/Entropy.md" {
		t.Errorf("the note it was made from has children %v, want the note just made", children)
	}
}

File: modules/apps/desktop/ui/src/creating.test.ts
/**
 * Making a note from the picture, and joining two that are already on it.
 *
 * What is checked here is what the vault cannot answer for: which seat the new
 * note writes about the one it was made from, where it is filed, and what
 * happens when the name is already taken.
 */
import { describe, expect, it } from 'vitest'

import { CREATABLE, UNTITLED, creating } from './creating'
import type { Core, Made, NewLink, NewNote } from './core'

const pathOf = (note: NewNote): string =>
  note.folder ? `${note.folder}/${note.title}.md` : `${note.title}.md`

/**
 * A core that keeps what it was asked to write and answers what a test told it
 * to, falling back on making the note.
 */
function fake(answers: Made[] = [], refusals: Made['refusal'][] = []) {
  const asked: NewNote[] = []
  const joined: { path: string; link: NewLink }[] = []
  const core = {
    create: async (note: NewNote): Promise<Made> => {
      asked.push(note)
      return answers.shift() ?? { path: pathOf(note), refusal: null }
    },
    join: async (path: string, link: NewLink): Promise<Made['refusal']> => {
      joined.push({ path, link })
      return refusals.shift() ?? null
    },
  } as unknown as Core
  return { core, asked, joined }
}

describe('making a note in a seat of another', () => {
  it('writes the note it was made from into it, in the seat facing the one asked for', async () => {
    const { core, asked } = fake()
    const made = await creating(core).make('Ontology.md', 'child')

    expect(made).toStrictEqual({ path: `${UNTITLED}.md`, title: UNTITLED })
    expect(asked).toStrictEqual([
      { title: UNTITLED, folder: '', links: [{ to: 'Ontology.md', seat: 'parent' }] },
    ])
  })

  it('makes a parent the new note is the child of', async () => {
    const { core, asked } = fake()
    await creating(core).make('Ontology.md', 'parent')

    expect(asked[0]?.links).toStrictEqual([{ to: 'Ontology.md', seat: 'child' }])
  })

  it('files it in the folder the note it was made from is in', async () => {
    const { core, asked } = fake()
    const made = await creating(core).make('physics/Ontology.md', 'child')

    expect(asked[0]?.folder).toBe('physics')
    expect(made?.path).toBe(`physics/${UNTITLED}.md`)
  })

  it('can make a note in every seat it offers', async () => {
    for (const seat of CREATABLE) {
      const { core, asked } = fake()
      expect(await creating(core).make('Ontology.md', seat)).not.toBeNull()
      expect(asked[0]?.links).toHaveLength(1)
    }
  })

  it('makes nothing in a seat no link writes', async () => {
    const { core, asked } = fake()

    expect(await creating(core).make('Ontology.md', 'sibling')).toBeNull()
    expect(asked).toStrictEqual([])
  })

  it('asks for the next name for as long as the vault says the last one is taken', async () => {
    const { core, asked } = fake([
      { path: '', refusal: 'occupied' },
      { path: '', refusal: 'occupied' },
    ])
    const making = creating(core)
    const made = await making.make('Ontology.md', 'child')

    expect(asked.map((note) => note.title)).toStrictEqual([
      UNTITLED,
      `${UNTITLED} 2`,
      `${UNTITLED} 3`,
    ])
    expect(made).toStrictEqual({ path: `${UNTITLED} 3.md`, title: `${UNTITLED} 3` })
    expect(making.said.value).toBe('')
  })

  it('says a refusal that is not a name already taken, and asks for nothing more', async () => {
    const { core, asked } = fake([{ path: '', refusal: 'notANote' }])
    const making = creating(core)

    expect(await making.make('Ontology.md', 'child')).toBeNull()
    expect(asked).toHaveLength(1)
    expect(making.said.value).not.toBe('')
  })

  it('says a core that could not be reached', async () => {
    const core = {
      create: async () => {
        throw new Error('the vault is out of reach')
      },
    } as unknown as Core
    const making = creating(core)

    expect(await making.make('Ontology.md', 'child')).toBeNull()
    expect(making.said.value).toContain('out of reach')
  })
})

describe('a note made on its own', () => {
  it('is filed at the top of the vault, joined to nothing', async () => {
    const { core, asked } = fake()
    const made = await creating(core).start()

    expect(made).toStrictEqual({ path: `${UNTITLED}.md`, title: UNTITLED })
    expect(asked).toStrictEqual([{ title: UNTITLED, folder: '', links: [] }])
  })

  it('takes the next free name, as a note made in a seat does', async () => {
    const { core } = fake([{ path: '', refusal: 'occupied' }])

    expect(await creating(core).start()).toStrictEqual({
      path: `${UNTITLED} 2.md`,
      title: `${UNTITLED} 2`,
    })
  })

  it('is nothing when the vault refused, and the refusal is said', async () => {
    const { core } = fake([{ path: '', refusal: 'unreadable' }])
    const making = creating(core)

    expect(await making.start()).toBeNull()
    expect(making.said.value).not.toBe('')
  })
})

describe('joining two notes that are both there', () => {
  it('writes the link in the note the gesture came from, in the seat it landed in', async () => {
    const { core, joined } = fake()
    const making = creating(core)

    expect(await making.join('Ontology.md', 'Entropy.md', 'child')).toBe(true)
    expect(joined).toStrictEqual([
      { path: 'Ontology.md', link: { to: 'Entropy.md', seat: 'child' } },
    ])
    expect(making.said.value).toBe('')
  })

  it('says why nothing was written', async () => {
    const { core } = fake([], ['unreadable'])
    const making = creating(core)

    expect(await making.join('Ontology.md', 'Entropy.md', 'jump')).toBe(false)
    expect(making.said.value).not.toBe('')
  })
})
